### PR TITLE
bpo-xxx: Fix potential crash in sqlite3 module when GC clears Cursor objects

### PR DESCRIPTION
**Pull request by erlend-aasland**

### CODE REVIEW

File: Modules/_sqlite/cursor.c [hunk 1]
```diff
@@ -121,6 +117,10 @@ cursor_dealloc(pysqlite_Cursor *self)
     if (self->in_weakreflist != NULL) {
         PyObject_ClearWeakRefs((PyObject*)self);
     }
+    /* Reset the statement if the user has not closed the cursor */
+    if (self->statement) {
```
Comment (pablogsal) [inline Modules/_sqlite/cursor.c:121]: This will always be NULL if the GC has collected the object in a cycle because you called `Py_CLEAR` in `tp_clear`.

**Comment by pablogsal** [comment]
This cannot crash in the same way the other one because this one is not dropping the GIL in the middle of tp_clear.

File: Modules/_sqlite/cursor.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c:121]: Correct. Thanks.

**Comment by erlend-aasland** [comment]
> This cannot crash in the same way the other one because this one is not dropping the GIL in the middle of tp_clear.

https://github.com/python/cpython/blob/fa106a685c1f199aca5be5c2d0277a14cc9057bd/Modules/_sqlite/statement.c#L374-L384

File: Modules/_sqlite/cursor.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c:121]: Maybe I'm missing something: `tp_clear` is called _after_ statement reset in `cursor_dealloc`.

Comment (pablogsal) [reply Modules/_sqlite/cursor.c:121]: > Maybe I'm missing something: `tp_clear` is called _after_ statement reset in `cursor_dealloc`.

The GC calls `tp_clear` *before* any objects gets deallocated

**Comment by pablogsal** [comment]
> `cursor_clear` may call `pysqlite_statement_reset` which may unlock GIL, no?

Actually, this is correct. Unfortunately I checking this wrong as I though `pysqlite_statement_reset` was part of libsqlite interface. Sigh.

File: Modules/_sqlite/cursor.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c:121]: So the explicit `tp_clear` in `cursor_dealloc` is just for corner cases?

**Pull request closed without merging**